src/hotspot/share/opto/escape.cpp

```diff
@@ -401,6 +416,239 @@ bool ConnectionGraph::compute_escape() {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. Input %d is nullable.", phi->_idx, i);)
+      return false;
+    }
+
+    // We are looking for at least one SR object in the merge
+    JavaObjectNode* ptn = unique_java_object(phi->in(i));
+    if (ptn != NULL && ptn->scalar_replaceable()) {
+      assert(ptn->ideal_node() != NULL && ptn->ideal_node()->is_Allocate(), "sanity");
+      Node* alloc = ptn->ideal_node()->as_Allocate();
+
+      if (PhaseMacroExpand::can_eliminate_allocation(_igvn, alloc->as_Allocate(), NULL, true)) {
+        found_sr_allocate = true;
+      } else {
+        ptn->set_scalar_replaceable(false);
+      }
+    }
+  }
+
+  NOT_PRODUCT(if (TraceReduceAllocationMerges && !found_sr_allocate) tty->print_cr("Can NOT reduce Phi %d. No SR Allocate as input.", phi->_idx);)
+  return found_sr_allocate;
+}
+
+// Check if we are able to untangle the merge. Right now we only reduce Phis
+// which are only used as debug information.
+bool ConnectionGraph::can_reduce_this_phi_users(PhiNode* phi) const {
+  for (DUIterator_Fast imax, i = phi->fast_outs(imax); i < imax; i++) {
+    Node* use = phi->fast_out(i);
+
+    if (use->is_SafePoint()) {
+      if (use->is_Call() && use->as_Call()->has_non_debug_use(phi)) {
+        NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. Call has non_debug_use().", phi->_idx);)
+        return false;
+      }
+    } else {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. One of the uses is: %d %s", phi->_idx, use->_idx, use->Name());)
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Returns true if: 1) It's profitable to reduce the merge, and 2) It The Phi is
+// only used in some certain code shapes. Check comments in
+// 'can_reduce_this_phi_inputs' and 'can_reduce_this_phi_users' for more
+// details.
+bool ConnectionGraph::can_reduce_this_phi(PointsToNode* var) const {
+  assert(var->escape_state() == PointsToNode::NoEscape, "Phi should be NoEscape");
+
+  PhiNode* phi = var->ideal_node()->as_Phi();
+  const Type* phi_t = _igvn->type(phi);
+  if (phi_t == NULL || phi_t->isa_instptr() == NULL || !phi_t->is_instptr()->klass_is_exact()) {
+    return false;
+  }
+
+  if (!can_reduce_this_phi_inputs(phi) || !can_reduce_this_phi_users(phi)) {
+    return false;
+  }
+
+  NOT_PRODUCT(if (TraceReduceAllocationMerges) { tty->print_cr("Can reduce this Phi during invocation %d: ", _invocation); phi->dump(); })
+  return true;
+}
+
+// This method will create SafePointScalarObjectNode for each combination of
+// scalar replaceable allocation in 'var' and SafePoint node in 'safepoints'.
+// Each SafePointScalarObjectNodes created here may describe multiple scalar
+// replaced objects - check detailed description in SafePointScalarObjectNode
+// class header.
+//
+// This method will set to 'NULL' entries in the Phi that are scalar replaceable.
```

Comment:
nit: "This method will set to 'NULL' entries in the Phi that are scalar replaceable." -> "This method will set entries in the Phi that are scalar replaceable to 'NULL'."